draft-irtf-cfrg-opaque.md

```diff
@@ -548,6 +552,11 @@ value is encoded as a byte string, specific to the AKE protocol in which OPAQUE
 is instantiated. These two options are defined as the `external` and `internal`
 modes, respectively. See {{envelope-modes}} for their specifications.
 
+The internal mode is RECOMMENDED. Applications can use the external mode if there
+are additional requirements for how private keys are generated, e.g., in the case
+of compliance, or if applications have pre-existing private keys they wish to
+register for use with OPAQUE.
```

Comment:
> Another argument for support for the external mode is the usage of cryptographic primitives for the AKE that don't support key pair creation/recovery from the OPRF output.

I can't see how this would be the case. Fundamentally, there must be some way for any key for any protocol to be derived from any sufficient amount of randomness, so it seems like internal mode will _always_ work.

Reply:
@hugokraw, what would be your argument for keeping external mode?

Reply:
isn't smaller packet sizes and generally smaller communication overhead an argument? imagine you have to run this over some low bandwidth radio connection that does afsk1200 for example...

Reply:
@stef I think external mode is more costly in terms of bandwidth, no? (Or maybe you meant a argument _against_ external mode, which I would agree with.)

Reply:
oh. i guess i'm mistaken, sorry. long time that i looked at the spec, i thought external refers to the keys being derived and not packed in the credentials structure (whatever that is called now)
note to self: read the spec before smartassing again...

Reply:
Both have advantages, their own trade-offs. IMHO, the internal mode is less exposed to implementation and usage errors.

Reply:
> @hugokraw, what would be your argument for keeping external mode?

We had a long discussion on this in the past. It is mathematically true that one can always run a key generation on a (pseudo) random seed. But different KG cases will have different complexity. For example, for RSA you would need to test numbers for primality, etc.  (there are some ways to simplify this in the case of RSA). The case where the external method is essential is if the key is imported and not generated at the time of key registration. I don't have a case in mind where this would be essential. Since this is not a security question I would defer to you guys on deciding this. If you remove it, I would still leave a note that applications with complicated KGs or those that are interested to import keys may decide to add the key to the envelope (although this mechanism is not specified here).